File: BoolSearch.py
'''
布尔序列二分查找工具类
任何二分查找问题都可以抽象为对Bool序列分界点的查找问题
查找[False...False,True...True]序列的第一个True，请使用right方法
查找[True...True,False...False]序列的最后一个True，请使用left方法
需要Python 3.10以上环境
by leetcode-cn.com/u/ak-bot
2022-3-26
'''
class BoolSearch:
    @staticmethod
    def right(l,r,func,param=None):
        if param:
            while l<r:
                mid=(l+r)//2
                if func(mid,param):
                    r=mid
                else:
                    l=mid+1
            return r
        else:
            while l<r:
                mid=(l+r)//2
                if func(mid):
                    r=mid
                else:
                    l=mid+1
            return r

    @staticmethod
    def left(l,r,func,param=None):
        if param:
            while l<r:
                mid=(l+r)//2+1
                if func(mid,param):
                    l=mid
                else:
                    r=mid-1
            return l
        else:
            while l<r:
                mid=(l+r)//2+1
                if func(mid):
                    l=mid
                else:
                    r=mid-1
            return l

File: BoolSearch.cpp
int bisect_right(int l,int r,bool (*pf)(int)){
    /*
    查找右侧第一个True
    对于一个形如[False...False,True...True]的序列
    此方法用于获取右侧第一个True的下标
    [l,r]是查找范围
    pf是一个函数，返回指定坐标x的布尔值
    */
    while(l<r){
        int mid = (l+r)/2;
        if((*pf)(mid))
            r=mid;
        else 
            l=mid+1;
        }
    return r;
}

int bisect_left(int l,int r,bool (*pf)(int)){
    /*
    查找左侧最后一个True
    对于一个形如[True...True,False...False]的序列
    此方法用于获取左侧最后一个True的下标
    [l,r]是查找范围
    pf是一个函数，返回指定坐标x的布尔值
    */
    while(l<r){
        int mid=(l+r)/2+1;
        if((*pf)(mid))
            l=mid;
        else
            r=mid-1;
    }
    return l;
}